Dockerfile

```diff
@@ -1,5 +1,40 @@
 # Build stage
-FROM docker.io/library/golang:1.21-alpine3.18 AS build-env
+FROM docker.io/library/node:20-alpine3.18 AS build-frontend
+
+ARG GITEA_VERSION
+
+# Build deps
+RUN apk --no-cache add \
+    build-base \
+    git \
+    && rm -rf /var/cache/apk/*
+
+# Setup repo
+WORKDIR /usr/src/code.gitea.io/gitea
+
+COPY Makefile .
+
+# Download NPM Packages
+COPY package.json .
+COPY package-lock.json .
+
+RUN make deps-frontend
+
+# Copy source files
+COPY ./webpack.config.js .
+COPY ./assets ./assets
+COPY ./public ./public
+COPY ./web_src ./web_src
+
+# Checkout version if set
+COPY ./.git ./.git
+RUN if [ -n "${GITEA_VERSION}" ]; then git checkout "${GITEA_VERSION}"; fi
```

Comment:
Thoughts on removing this line, can see it has been untouched in this file for 5 years so guessing it might be a part of a workflow?

Disadavntage of including these lines is that any time a maintainer commits to git, they will be re-running the `COPY ./.git ./.git` step. Currently this happend transparently.

To replicate the functionality you can use this feature of [docker build](https://docs.docker.com/engine/reference/commandline/build/)

```bash
docker build https://github.com/go-gitea/gitea.git#v1.20.5
```

Reply:
This is due to the makefile using git calls to determine the version information. This could (and tbh should) be pulled out for the reasons you mentioned that whenever git repo changes it invalidates the cache (even if you are attempting to build the same tag). We do some magic in the makefile to calculate the version, and this logic is used for more than just docker (eg xgo crossarch builds) which means it's kinda load bearing and changes to it need a closer eye when reviewing.

Reply:
Looking at the failed build, the `.git` folder isn't present during CI runs - I think this is more to do with the dyrun workflow needing to do a full checkout once it has detected a file has changed?

In the Dockerfile `GITEA_VERSION` is an [`ARG`](https://docs.docker.com/engine/reference/builder/#using-arg-variables), from what I can tell these are only accessible within the Dockerfile, and aren't in the build as environment variables, so should be safe to take out?

I think in the GitHub release workflows, it is using the `.git` directory to determine version information.

Reply:
I think if we were to go down the route of not relying on `.git` for Docker builds, we may want to go down the route of something like....

```dockerfile
ARG GITEA_VERSION
ENV GITEA_VERSION=${GITEA_VERSION:-develop}
RUN echo "${GITEA_VERSION}" > VERSION
```

This way, when developing you will get a nominal "develop" value for your Gitea version, but when compiling for release, it can be specified as a build argument?